torch/csrc/jit/codegen/cuda/disjoint_set.h

```diff
@@ -0,0 +1,127 @@
+#pragma once
+
+#include <c10/util/Exception.h>
+
+#include <unordered_map>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//! Container class DisjointSet models equivalence relationships
+//!
+//! Each instance of this class keeps a set of equivalent classes
+//! DisjointSet::join(a,b) makes the full class of a and b equivalent
+//! DisjointSet::areEqual(a,b) checks if a and b belong same class
+//!
+//! \note The template type T is assumed to be hashable
+template <typename T>
+class DisjointSet {
+ public:
+  DisjointSet() = default;
+
+  //! Joins the equivalent class that a and b belong to
+  //! areEqual(a',b') will be true for each a'=a and b'=b
+  //!
+  //! \param a An element from a equivalent class
+  //!          will create a new equivalent class if a does
+  //!          not belong to any
+  //! \param b An element from another equivalent class
+  //!          will create a new equivalent class if b does
+  //!          not belong to any
+  void join(T a, T b) {
+    // cases where either of the quiv class doesn't exist
+    if (!entry_map.count(a) && !entry_map.count(b)) {
+      createPoint(a);
+      entry_map[b] = fixedPoint(a);
+    } else if (!entry_map.count(a)) {
+      entry_map[a] = fixedPoint(b);
+    } else if (!entry_map.count(b)) {
+      entry_map[b] = fixedPoint(a);
+    } else {
+      // case where both equiv classes exist and need to join
+      const int i0 = fixedPoint(a);
+      const int i1 = fixedPoint(b);
+      int new_parent = 0;
+      int new_child = 0;
+
+      // Either order here is correct but joining larger class to smaller class
+      // tend to be faster
+      std::tie(new_parent, new_child) = (weights[i0] < weights[i1])
+          ? std::make_pair(i0, i1)
+          : std::make_pair(i1, i0);
+      weights[new_parent] += weights[new_child];
+      set_map[new_child] = new_parent;
+    }
+  }
+
+  //! Checks if a and b belong to the same equivalent class
+  //!
+  //! \param a An element from a equivalent class
+  //! \param b An element from another equivalent class
+  //! \returns Boolean value representing if a and b are
+  //!          recorded to be in the same equivalent class
+  //!          will return false if any of a or b doesn't
+  //!          have an equivalent class recorded
+  bool areEquivalent(T a, T b) const {
+    if (!entry_map.count(a) || !entry_map.count(b)) {
+      return false;
+    }
+    return fixedPoint(a) == fixedPoint(b);
+  }
+
+ private:
+  // Internal fixed point implementation:
+  //  Returns the equivalent class that e belongs to
+  int getFixedPointForClass(int e) const {
```

Comment:
@shmsong I changed the name of this function as it resulted in a name conflict when `T` is `int`.